### ISSUE TITLE
Clarify when words hold particular definitions

### ISSUE THREAD

**Issue by aidantwoods** [commonmark/CommonMark#476]
Proposal that words which hold particular definitions always hyperlink to their definition, or have some other way of distinguishing them.

Example of ambiguity:
> [Autolink](http://spec.commonmark.org/0.27/#autolink)s are absolute URIs and email addresses inside `<` and `>`. They are parsed as links, with the URL or email address as the link label.

Does the use of the words "link" and "links" correspond to HTML element `<a>`, or the definition (*)?

(*)
> A link contains [link text](http://spec.commonmark.org/0.27/#link-text) (the visible text), a [link destination](http://spec.commonmark.org/0.27/#link-destination) (the URI that is the link destination), and optionally a [link title](http://spec.commonmark.org/0.27/#link-title). There are two basic kinds of links in Markdown. In [inline links](http://spec.commonmark.org/0.27/#inline-link) the destination and title are given immediately after the link text. In [reference links](http://spec.commonmark.org/0.27/#reference-link) the destination and title are defined elsewhere in the document.

From the definition, it would appear that this particular definition only applies to *inline links* and *reference links*. However this section is not linked to when used

e.g. in the definition for *link text*
> Links may not contain other links, at any level of nesting. If multiple otherwise valid link definitions appear nested inside each other, the inner-most definition is used.

To repeat the initial question: Does the use of the words "link" and "links" correspond to HTML element `<a>`, or the definition (*)?

Which leads to another question: Can *link text* contain an *autolink*?

The definition of an *autolink* states that it **is** a "link" in the same way that *link text* states that it may not contain any "links".

---

For the purposes of demonstrating how this ambiguity may be resolved (and generally because it would be incomprehensible otherwise), I have used emphasis in the above to indicate when a word holds a definition defined in the CommonMark document. However in blockquotes, hyperlinks have been preserved from the document and no words are emphasised for the purpose of accuracy to the original text.

**Comment by aidantwoods** [comment]
Reference parser seems to [imply](http://spec.commonmark.org/dingus/?text=%5Bfoo%3Chttp%3A%2F%2Fgoogle.com%3E%5D(bar)%0A%0A---%0A%0A%5Bfoo%5Bbar%5D(boo)%5D(baz)%0A%0A---%0A%0A%5Bbar%5D%3A%20bop%0A%0A%5Bfoo%5Bbar%5D%5D(baz)%0A) that use of the word "links" in the *autolink* definition means `<a>` while use of the same word in *link text* definition means either a *reference link* or an *inline link*.

**Comment by aidantwoods** [comment]
I can open another bug for the particular example issue if you like, this is more of a meta-bug for a problem that occurs throughout the spec.

Another example, in the section about [tabs](http://spec.commonmark.org/0.27/#tabs):

> Tabs in lines are not expanded to [spaces](http://spec.commonmark.org/0.27/#space). However, in contexts where whitespace helps to define block structure, tabs behave as if they were replaced by spaces with a tab stop of 4 characters.

The first instance of *spaces* is linked to, but "whitespace" is not. What does whitespace mean here? Also, having a contextually overloaded definition is very ambiguous on its own, and especially when "block structure" isn't formally defined. What qualifies a character to "help defined block structure"? Every non-whitespace character prevents a paragraph breaking, and a line of all whitespace characters breaks a paragraph. Does every character then "help defined block structure"? Do all mentions of the word "spaces" therefore hold this property?

I would propose for this (again separate issue) introducing the term **structural space**.

Such that a ***structural space*** is defined to be, either
* A *space*, or
* A tab, as if it were replaced by spaces with a tab stop of 4 characters.

Then when definitions actually mean this special type of space, they can use the term *structural space* and reference it in some way.


---

Prior to solving this example, we should decide on whether to always hyperlink to the definition, or to only have a single special definition for each word, and just use emphasis whenever we want to mean to special definition, e.g.

e.g.  when "link" means a *inline link* or or *reference link* we use *link* so that we know it does not mean `<a>`.

---

This only really causes ambiguity when CommonMark defined words have different definitions from their dictionary (or at least web-context) definitions, so those are probably the only cases that'll really need this. But maybe as good practice, we just always leave some way of knowing whether these words hold the CommonMark defined definition, or the "standard" definition.


As an example that does this kind of thing with a lot of hyperlinking, take a look at w3c documentation, e.g. https://www.w3.org/TR/referrer-policy/#referrer-policy-no-referrer

**Comment by aidantwoods** [comment]
Lastly, I'd be happy to help with rewriting the spec to include references where special definitions of words are used – but let's have a discussion about how we want to proceed on a formatting level first

**Comment by jgm** [comment]
You're right that there are some problematic ambiguities around "link."  Perhaps you should close this and open a separate issue devoted just to that (or change the title of this one).

On the general question, I think it would be desirable to hyperlink words to their definitions whenever possible.  (I didn't do this from the very beginning, but added the links later, so not everything is linked.)

On tabs:  clarifying the language here is already on the list of things that should happen before 1.0, on the forum.  If there's not an issue for that here, opening that would be welcome.

**Comment by jgm** [comment]
+++ Aidan Woods [May 27 17 14:12 ]:
>   Such that a structural space is defined to be, either
>     * A space, or
>     * A tab, as if it were replaced by spaces with a tab stop of 4
>       characters.

Tricky to word this right, because we want to give a meaning
to e.g. "two structural spaces."  The two structural spaces
might both be part of a tab, or one might be part of one
tab, one part of another, or one might be a space and the
other part of a tab, etc.

**Comment by jgm** [comment]
+++ Aidan Woods [May 27 17 14:17 ]:
>   Lastly, I'd be happy to help with rewriting the spec to include
>   references where special definitions of words are used – but let's have
>   a discussion about how we want to proceed on a formatting level first

Excellent, this help would be very much welcome!

**Comment by aidantwoods** [comment]
> You're right that there are some problematic ambiguities around "link."  Perhaps you should close this and open a separate issue devoted just to that (or change the title of this one).

I'll open another issue for that, shall I leave this open to track the hyperlinking progress though (mentions of the ambiguities around "link" were mainly to demonstrate why the hyperlinking would be useful in the spec – though hard to really give a good argument without delving into the example in full 😉). The intent of this issue was for general definition clarity though.

---

> On tabs:  clarifying the language here is already on the list of things that should happen before 1.0, on the forum.  If there's not an issue for that here, opening that would be welcome.

Awesome! There is mention of this in https://github.com/jgm/CommonMark/issues/464, though only to a specific block definition – happy to open another to propose some from of structural spaces though, let me know!

**Comment by aidantwoods** [comment]
> Tricky to word this right, because we want to give a meaning
to e.g. "two structural spaces."  The two structural spaces
might both be part of a tab, or one might be part of one
tab, one part of another, or one might be a space and the
other part of a tab, etc.

You're right about that, perhaps something like this would be a little better:

Such that a ***structural space*** is defined to be, either
* A *space*, or
* Part of a tab, as if the tab were replaced by spaces with a tab stop of 4 characters, where each of these replacement spaces is a *structural space*.

I'm sure there is a nicer way to word that though.

**Comment by aidantwoods** [comment]
@jgm 
> Excellent, this help would be very much welcome!

Is there a primer on the spec format anywhere. Looks to me like `#` headings are auto id-'d, and `[foo](@)` is a definition anchor, and any anchor point is referenceable with `[bar]`. Little confused as to how this works though

```
A [whitespace character](@) is a space
(`U+0020`), tab (`U+0009`), newline (`U+000A`), line tabulation (`U+000B`),
form feed (`U+000C`), or carriage return (`U+000D`).

[Whitespace](@) is a sequence of one or more [whitespace
characters].
```

Does the trailing `s` not matter in `[whitespace characters]`?

**Comment by jgm** [comment]
If you look in tools/make_spec.lua, you'll see the script
that adds these custom linking features.  It uses filters
for lcmark, a lua wrapper of cmark.

This part adds a reference for the pluralized form whenever
you define one:

      if not refs[label .. 's'] then
        -- plural too
        refs[label .. 's'] = ident
      end

Ideally this should be documented of course!
